datafusion/optimizer/src/reduce_cross_join.rs

```diff
@@ -0,0 +1,442 @@
+                        join_constraint: JoinConstraint::On,
+                        on: join_keys,
+                        filter: None,
+                        schema: cross_join.schema.clone(),
+                        null_equals_null: false,
+                    }))
+                }
+            }
+            _ => {
+                let expr = plan.expressions();
+
+                // apply the optimization to all inputs of the plan
+                let inputs = plan.inputs();
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| {
+                        reduce_cross_join(
+                            _optimizer,
+                            plan,
+                            possible_join_keys,
+                            all_join_keys,
+                            _optimizer_config,
+                        )
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                from_plan(plan, &expr, &new_inputs)
+            }
+    }
+}
+
+fn intersect(
+    accum: &mut Vec<(Column, Column)>,
+    vec1: &[(Column, Column)],
+    vec2: &[(Column, Column)],
+) -> Result<()>  {
+
+    for x1 in vec1.iter() {
+        for x2 in vec2.iter() {
+            if x1.0 == x2.0 && x1.1 == x2.1
+                || x1.1 == x2.0 && x1.0 == x2.1
+            {
+                accum.push((x1.0.clone(), x1.1.clone()));
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Extract join keys from a WHERE clause
+fn extract_possible_join_keys(
+    expr: &Expr,
+    accum: &mut Vec<(Column, Column)>,
+) -> Result<()> {
+    match expr {
+        Expr::BinaryExpr { left, op, right } => match op {
+            Operator::Eq => match (left.as_ref(), right.as_ref()) {
+                (Expr::Column(l), Expr::Column(r)) => {
+                    // Ensure that we don't add the same Join keys multiple times
+                    if !(accum.contains(&(l.clone(), r.clone())) ||
+                        accum.contains(&(r.clone(), l.clone()))) {
+                        accum.push((l.clone(), r.clone()));
+                    }
+                    Ok(())
+                }
+                _ => Ok(()),
+            },
+            Operator::And => {
+                extract_possible_join_keys(left, accum)?;
+                extract_possible_join_keys(right, accum)
+            },
+            // Fix for issue#78 join predicates from inside of OR expr also pulled up properly.
+            Operator::Or => {
+                let mut left_join_keys = vec![];
+                let mut right_join_keys = vec![];
+
+                extract_possible_join_keys(left, &mut left_join_keys)?;
+                extract_possible_join_keys(right, &mut right_join_keys)?;
+
+                intersect( accum, &left_join_keys, &right_join_keys)
+            }
+            _ => Ok(()),
+        },
+        _ => Ok(()),
+    }
+}
+
+/// Remove join expressions from a filter expression
```

Comment:
Additional documentation about return values would help here, e.g. what do `Some()` and `None` mean in context?

Reply:
ok